src/NonLinearProgram/nlp_utilities.jl

```diff
@@ -0,0 +1,556 @@
+)
+    evaluator = model.cache.evaluator
+    jacobian_sparsity = MOI.jacobian_structure(evaluator)
+    I = [i for (i, _) in jacobian_sparsity]
+    J = [j for (_, j) in jacobian_sparsity]
+    V = zeros(length(jacobian_sparsity))
+    MOI.eval_constraint_jacobian(evaluator, V, model.x)
+    A = SparseArrays.sparse(I, J, V, length(rows), length(model.x))
+    return A
+end
+
+"""
+    compute_optimal_hess_jac(evaluator::MOI.Nonlinear.Evaluator, rows::Vector{JuMP.ConstraintRef}, x::Vector{JuMP.VariableRef})
+
+Compute the optimal Hessian of the Lagrangian and Jacobian of the constraints.
+"""
+function compute_optimal_hess_jac(
+    model::Model,
+    rows::Vector{MOI.Nonlinear.ConstraintIndex},
+)
+    hessian = compute_optimal_hessian(model, rows)
+    jacobian = compute_optimal_jacobian(model, rows)
+
+    return hessian, jacobian
+end
+
+"""
+    create_evaluator(form::Form)
+
+Create the evaluator for the NLP.
+"""
+function create_evaluator(form::Form)
+    nlp = form.model
+    backend = MOI.Nonlinear.SparseReverseMode()
+    evaluator = MOI.Nonlinear.Evaluator(
+        nlp,
+        backend,
+        MOI.VariableIndex.(1:form.num_variables),
+    )
+    MOI.initialize(evaluator, [:Hess, :Jac])
+    return evaluator
+end
+
+"""
+    is_less_inequality(con::MOI.ConstraintIndex{F, S}) where {F, S}
+
+Check if the constraint is a less than inequality.
+"""
+function is_less_inequality(
+    ::MOI.ConstraintIndex{F,S},
+) where {
+    F<:Union{
+        MOI.ScalarNonlinearFunction,
+        MOI.ScalarQuadraticFunction{Float64},
+        MOI.ScalarAffineFunction{Float64},
+    },
+    S<:MOI.LessThan,
+}
+    return true
+end
+
+function is_less_inequality(::MOI.ConstraintIndex{F,S}) where {F,S}
+    return false
+end
+
+function is_greater_inequality(::MOI.ConstraintIndex{F,S}) where {F,S}
+    return false
+end
+
+"""
+    is_greater_inequality(con::MOI.ConstraintIndex{F, S}) where {F, S}
+
+Check if the constraint is a greater than inequality.
+"""
+function is_greater_inequality(
+    ::MOI.ConstraintIndex{F,S},
+) where {
+    F<:Union{
+        MOI.ScalarNonlinearFunction,
+        MOI.ScalarQuadraticFunction{Float64},
+        MOI.ScalarAffineFunction{Float64},
+    },
+    S<:MOI.GreaterThan,
+}
+    return true
+end
+
+"""
+    find_inequealities(cons::Vector{JuMP.ConstraintRef})
+
+Find the indices of the inequality constraints.
+"""
+function find_inequealities(model::Form)
+    num_cons = length(model.list_of_constraint)
+    leq_locations = zeros(num_cons)
+    geq_locations = zeros(num_cons)
+    for con in keys(model.list_of_constraint)
+        if is_less_inequality(con)
+            leq_locations[model.constraints_2_nlp_index[con].value] = true
+        end
+        if is_greater_inequality(con)
+            geq_locations[model.constraints_2_nlp_index[con].value] = true
+        end
+    end
+    return findall(x -> x == 1, leq_locations),
+    findall(x -> x == 1, geq_locations)
```

Comment:
I would put the two `findall`on the same line

Reply:
This was made by formatter. I believe we need it to pass the format test